File: 847.shortest-path-visiting-all-nodes.ts
/*
 * @lc app=leetcode id=847 lang=typescript
 *
 * [847] Shortest Path Visiting All Nodes
 */

// @lc code=start
class NodeState {
  public index: number;
  public mask: number;
  public steps: number;

  constructor(index: number, mask: number, steps: number) {
    this.index = index;
    this.mask = mask;
    this.steps = steps;
  }
}

function hashMask(mask: number, index: number): number {
  return (mask << 4) + index;
}

function shortestPathLength(graph: number[][]): number {
  const visited = {};
  const targetState = (1 << graph.length) - 1;

  const queue: NodeState[] = [];
  for (let i = 0; i < graph.length; i++) {
    queue.push(new NodeState(i, 1 << i, 0));
  }

  while (queue.length > 0) {
    let { index, mask, steps } = queue.shift();

    if (mask == targetState) {
      return steps;
    }

    for (let n of graph[index]) {
      let newMask = mask | (1 << n);
      if (!visited[hashMask(newMask, n)]) {
        queue.push(new NodeState(n, newMask, steps + 1));
        visited[hashMask(newMask, n)] = true;
      }
    }
  }

  return -1;
}
// @lc code=end

File: 847.shortest-path-visiting-all-nodes.rb
#
# @lc app=leetcode id=847 lang=ruby
#
# [847] Shortest Path Visiting All Nodes
#

# @lc code=start
# @param {Integer[][]} graph
# @return {Integer}
def hash_mask(mask, index)
  (mask << 4) + index
end

def shortest_path_length(graph)
  node_state = Struct.new(:index, :mask, :steps)

  visited = {}
  target_state = (1 << graph.length) - 1

  queue = graph.each_with_index.map do |neighbors, index|
    node_state.new(index, 1 << index, 0)
  end

  until queue.empty?
    s = queue.shift
    index = s.index
    mask = s.mask
    steps = s.steps

    return steps if mask == target_state

    graph[index].each do |neighbor|
      new_mask = mask | (1 << neighbor)
      if !visited[hash_mask(new_mask, neighbor)]
        queue.push(node_state.new(neighbor, new_mask, steps + 1))
        visited[hash_mask(new_mask, neighbor)] = true
      end
    end

  end

  -1
end
# @lc code=end
